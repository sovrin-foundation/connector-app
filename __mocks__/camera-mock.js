// @flow
import React from 'react'
import { View } from 'react-native'

const constants = {
  Aspect: {},
  BarCodeType: {},
  Type: {},
  CaptureMode: {},
  CaptureTarget: {},
  CaptureQuality: {},
  Orientation: {},
  FlashMode: {},
  TorchMode: {},
}

// We don't need to type check external library
// and we don't need types for mock test stub
class MockCamera extends React.Component<any, void> {
  static constants = constants
  static checkVideoAuthorizationStatus = jest.fn(() => Promise.resolve(true))

  render() {
    return <View>{this.props.children}</View>
  }
}

MockCamera.constants = constants

export default MockCamera
