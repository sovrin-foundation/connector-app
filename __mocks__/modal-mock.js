// @flow
import React from 'react'

const MockModal = (props: any) =>
  React.createElement('Modal', props, props.children)

export default MockModal
