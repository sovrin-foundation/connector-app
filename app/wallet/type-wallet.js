// @flow
import type { ReactNavigation } from '../common/type-common'
import type { CustomError } from '../common/type-common'
import type { IsValid } from '../components/input-control/type-input-control'
import type { CredentialOfferModalStatus } from '../claim-offer/type-claim-offer'

export type BackupWalletProps = {
  backup: BackupInfo,
  render: (backup: string, backupWallet: () => void) => void,
  walletBackup: () => void,
} & ReactNavigation

export type BackupInfo = {
  latest: ?string,
  error: ?CustomError,
  backupPath: ?string,
} & StoreStatus

export const BACKUP_WALLET = 'BACKUP_WALLET'

export const BACKUP_WALLET_FAIL = 'BACKUP_WALLET_FAIL'

export const BACKUP_WALLET_SUCCESS = 'BACKUP_WALLET_SUCCESS'

export const GET_WALLET_ENCRYPTION_KEY = 'GET_WALLET_ENCRYPTION_KEY'

export const BACKUP_WALLET_PATH = 'BACKUP_WALLET_PATH'

export const SHARE_WALLET_BACKUP = 'SHARE_WALLET_BACKUP'

export const ERROR_BACKUP_WALLET = {
  code: 'WB-001',
  message: 'Error while backing up wallet',
}

export const ERROR_BACKUP_WALLET_SHARE = {
  code: 'WB-002',
  message: 'Error while sharing zipped backup',
}

export type WalletTabSendDetailsProps = {
  tokenAmount: string,
  tokenSentStatus: $Keys<typeof STORE_STATUS>,
  sendTokens: (
    tokenAmount: string,
    recipientAddress: string
  ) => SendTokensAction,
} & ReactNavigation

export type BackupWalletAction = {
  type: typeof BACKUP_WALLET,
  data: BackupInfo,
}

export type ShareBackupAction = {
  type: typeof SHARE_WALLET_BACKUP,
  data: BackupInfo,
}

export type WalletBalanceProps = {
  walletBalance: string,
  refreshWalletBalance: () => {},
  render: Function,
}

export type WalletSendAmountState = {
  text: string,
}

export type WalletSendAmountProps = {
  screenProps: ReactNavigation,
  selectTokenAmount: (tokenAmount: string) => {},
  walletBalance: string,
  paymentStatus: $Keys<typeof STORE_STATUS>,
} & ReactNavigation

export type WalletProps = {} & ReactNavigation

export type WalletPayload = {
  walletPath: string,
  encryptionKey: string,
}

export type WalletTabReceiveProps = {
  walletAddresses: Array<string>,
  refreshWalletAddresses: () => {},
  promptBackupBanner: (showBackup: boolean) => {},
  addressStatus: $Keys<typeof STORE_STATUS>,
  currentScreen: string,
}

export type WalletTabReceiveState = {
  copyButtonText: string,
}

export type WalletSendPaymentData = {
  paymentTo: string,
  paymentFor?: string,
}

export type WalletTabSendDetailsState = {
  showPaymentAddress: boolean,
  isPaymentAddressValid: IsValid,
  credentialOfferModalStatus: CredentialOfferModalStatus,
}

export type WalletHistoryProps = {
  walletHistory: WalletHistory,
  refreshWalletHistory: () => {},
}

export type WalletTabsProps = {} & ReactNavigation

export type WalletHistoryEvent = {
  id: string,
  senderName?: string,
  senderAddress: string,
  action: string,
  tokenAmount: string,
  timeStamp: string,
}

export type WalletBalance = {
  data: string,
} & StoreStatus &
  StoreError

export type WalletAddresses = {
  data: Array<string>,
} & StoreStatus &
  StoreError

export type WalletHistory = {
  transactions: WalletHistoryEvent[],
} & StoreStatus &
  StoreError

export type Payment = {
  tokenAmount: string,
} & StoreStatus &
  StoreError

export const ERROR_LOADING_WALLET = {
  code: 'W001',
  message: 'Error while loading wallet data',
}

export const ERROR_REFRESHING_WALLET_BALANCE = {
  code: 'W002',
  message: 'Error while refreshing wallet balance',
}

export const ERROR_REFRESHING_WALLET_ADDRESSES = {
  code: 'W003',
  message: 'Error while refreshing wallet addresses',
}

export const ERROR_REFRESHING_WALLET_HISTORY = {
  code: 'W004',
  message: 'Error while refreshing wallet history',
}

export const ERROR_LOADING_WALLET_BALANCE = {
  code: 'W005',
  message: 'Error while loading wallet balance',
}

export const ERROR_LOADING_WALLET_ADDRESSES = {
  code: 'W006',
  message: 'Error while loading wallet addresses',
}

export const ERROR_LOADING_WALLET_HISTORY = {
  code: 'W007',
  message: 'Error while loading wallet history',
}

export const ERROR_SENDING_TOKENS = {
  code: 'W008',
  message: 'Error sending tokens',
}

export const ERROR_SENDING_TOKENS_WITH_FEES = {
  code: 'W009',
  message:
    'Error sending tokens while transfer fees is greater than or equal to transfer amount',
}

export const STORE_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
}

export type StoreError = { error: ?CustomError }
export type StoreStatus = { status: $Keys<typeof STORE_STATUS> }

export type WalletStore = {
  walletBalance: WalletBalance,
  walletAddresses: WalletAddresses,
  walletHistory: WalletHistory,
  backup: BackupInfo,
  payment: Payment,
}

export type HydrateWalletBalanceFailAction = {
  type: typeof HYDRATE_WALLET_BALANCE_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletAddressesFailAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type HydrateWalletHistoryFailAction = {
  type: typeof HYDRATE_WALLET_HISTORY_FAIL,
  error: CustomError,
  status: StoreStatus,
}

export type SendTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
  recipientWalletAddress: string,
}

export type SelectTokensAction = {
  type: typeof SEND_TOKENS,
  tokenAmount: string,
}

export const HYDRATE_WALLET_BALANCE_FAIL = 'HYDRATE_WALLET_BALANCE_FAIL'
export const HYDRATE_WALLET_ADDRESSES_FAIL = 'HYDRATE_WALLET_ADDRESSES_FAIL'
export const HYDRATE_WALLET_HISTORY_FAIL = 'HYDRATE_WALLET_HISTORY_FAIL'
export const REFRESH_WALLET_BALANCE_FAIL = 'REFRESH_WALLET_BALANCE_FAIL'
export const REFRESH_WALLET_ADDRESSES_FAIL = 'REFRESH_WALLET_ADDRESSES_FAIL'
export const REFRESH_WALLET_HISTORY_FAIL = 'REFRESH_WALLET_HISTORY_FAIL'
export const HYDRATE_WALLET_BALANCE = 'HYDRATE_WALLET_BALANCE'
export const HYDRATE_WALLET_ADDRESSES = 'HYDRATE_WALLET_ADDRESSES'
export const HYDRATE_WALLET_HISTORY = 'HYDRATE_WALLET_HISTORY'
export const REFRESH_WALLET = 'REFRESH_WALLET'
export const REFRESH_WALLET_ADDRESSES = 'REFRESH_WALLET_ADDRESSES'
export const REFRESH_WALLET_HISTORY = 'REFRESH_WALLET_HISTORY'
export const REFRESH_WALLET_BALANCE = 'REFRESH_WALLET_BALANCE'
export const WALLET_BALANCE_REFRESHED = 'WALLET_BALANCE_REFRESHED'
export const WALLET_ADDRESSES_REFRESHED = 'WALLET_ADDRESSES_REFRESHED'
export const WALLET_HISTORY_REFRESHED = 'WALLET_HISTORY_REFRESHED'
export const SEND_TOKENS = 'SEND_TOKENS'
export const SEND_TOKENS_FAIL = 'SEND_TOKENS_FAIL'
export const TOKEN_SENT_SUCCESS = 'TOKEN_SENT_SUCCESS'
export const SELECT_TOKEN_AMOUNT = 'SELECT_TOKEN_AMOUNT'
export const PROMPT_WALLET_BACKUP_BANNER = 'PROMPT_WALLET_BACKUP_BANNER'
export const WALLET_ADDRESSES_FETCH_START = 'WALLET_ADDRESSES_FETCH_START'

export type PromptBackupBannerAction = {
  type: typeof PROMPT_WALLET_BACKUP_BANNER,
  showBanner: boolean,
}

export type HydrateWalletBalanceData = string

export type HydrateWalletAddressesData = Array<string>

export type HydrateWalletHistoryTransactions = WalletHistoryEvent[]

export type RefreshWalletHistoryAction = {
  type: typeof REFRESH_WALLET_HISTORY,
}

export type WalletBalanceRefreshedAction = {
  type: typeof WALLET_BALANCE_REFRESHED,
  walletBalance: WalletBalance,
}

export type WalletAddressesRefreshedAction = {
  type: typeof WALLET_ADDRESSES_REFRESHED,
  walletAddresses: WalletAddresses,
}

export type WalletHistoryRefreshedAction = {
  type: typeof WALLET_HISTORY_REFRESHED,
  walletHistory: WalletHistory,
}

export type RefreshWalletAddressesAction = {
  type: typeof REFRESH_WALLET_ADDRESSES,
}

export type RefreshWalletBalanceAction = {
  type: typeof REFRESH_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletBalanceAction = {
  type: typeof HYDRATE_WALLET_BALANCE,
  walletBalance: WalletBalance,
}

export type HydrateWalletAddressesAction = {
  type: typeof HYDRATE_WALLET_ADDRESSES,
  walletAddresses: WalletAddresses,
}

export type HydrateWalletHistoryAction = {
  type: typeof HYDRATE_WALLET_HISTORY,
  walletHistory: WalletHistory,
}

export type TokenSentSuccessAction = {
  type: typeof TOKEN_SENT_SUCCESS,
  payment: Payment,
}

export type WalletStoreAction =
  | HydrateWalletBalanceFailAction
  | HydrateWalletAddressesFailAction
  | HydrateWalletHistoryFailAction
  | RefreshWalletBalanceAction
  | RefreshWalletAddressesAction
  | RefreshWalletHistoryAction
  | WalletHistoryRefreshedAction
  | WalletBalanceRefreshedAction
  | WalletAddressesRefreshedAction
  | HydrateWalletBalanceAction
  | HydrateWalletAddressesAction
  | HydrateWalletHistoryAction
  | BackupWalletAction
  | SendTokensAction
  | TokenSentSuccessAction
  | SelectTokensAction
