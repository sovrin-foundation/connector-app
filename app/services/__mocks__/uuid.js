// @flow
import { defaultUUID } from '../../../__mocks__/static-data'

export function uuid() {
  return defaultUUID
}
