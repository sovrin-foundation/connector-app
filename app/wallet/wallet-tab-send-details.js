// @flow
import React, { Component } from 'react'
import { StyleSheet, Keyboard, Platform } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { createStackNavigator, NavigationActions } from 'react-navigation'
import { scale } from 'react-native-size-matters'
import {
  Container,
  CustomText,
  CustomView,
  CustomButton,
  CustomHeader,
  Icon,
  Loader,
} from '../components'
import { primaryHeaderStyles } from '../components/layout/header-styles'
import {
  OFFSET_1X,
  OFFSET_2X,
  OFFSET_3X,
  OFFSET_6X,
  OFFSET_7X,
  color,
  HAIRLINE_WIDTH,
  whiteSmokeSecondary,
} from '../common/styles/constant'
import ControlInput from '../components/input-control/input-control'
import type {
  WalletSendPaymentData,
  WalletTabSendDetailsProps,
  WalletTabSendDetailsState,
} from './type-wallet'
import { formatNumbers } from '../components/text'
import { receiveTabRoute } from '../common'
import { walletAddresses } from '../../__mocks__/static-data'
import {
  SEND_TOKENS_TO_PAYMENT_ADDRESS,
  FOR_SEND_DETAILS_TEST_ID,
  TO_SEND_DETAILS_TEST_ID,
  FONT_SIZE_MAPPING,
} from './wallet-constants'
import { sendTokens } from '../wallet/wallet-store'
import { getWalletAddresses, getTokenAmount } from '../store/store-selector'
import type { Store } from '../store/type-store'
import { STORE_STATUS } from './type-wallet'
import { CREDENTIAL_OFFER_MODAL_STATUS } from '../claim-offer/type-claim-offer'
import Modal from 'react-native-modal'
import PaymentFailureModal from './payment-failure-modal'
import CredentialOfferModal from './credential-offer-modal'

export class WalletTabSendDetails extends Component<
  WalletTabSendDetailsProps,
  WalletTabSendDetailsState
> {
  static navigationOptions = ({ navigation, screenProps }) => ({
    header: (
      <CustomHeader flatHeader backgroundColor={whiteSmokeSecondary}>
        <Icon
          testID={'back-arrow'}
          iconStyle={[styles.headerLeft]}
          src={require('../images/icon_backArrow.png')}
          resizeMode="contain"
          onPress={() => navigation.goBack(null)}
          small
        />

        <CustomText
          style={[{ color: color.bg.seventh.font.sixth }]}
          transparentBg
          h5
        >
          Pay Tokens
        </CustomText>

        <CustomText
          uppercase
          bold
          quinaryText
          transparentBg
          h5
          style={[navigation.state.params.isValid ? {} : styles.disabledText]}
          testID={SEND_TOKENS_TO_PAYMENT_ADDRESS}
          onPress={() => navigation.state.params.onSendTokens()}
        >
          Send
        </CustomText>
      </CustomHeader>
    ),
  })

  state = {
    showPaymentAddress: false,
    isPaymentAddressValid: 'IDLE',
    credentialOfferModalStatus: CREDENTIAL_OFFER_MODAL_STATUS.NONE,
  }

  paymentData: WalletSendPaymentData = {
    paymentTo: '',
    paymentFor: '',
  }

  componentDidUpdate(prevProps: WalletTabSendDetailsProps) {
    if (this.props.tokenSentStatus !== prevProps.tokenSentStatus) {
      if (this.props.tokenSentStatus === STORE_STATUS.SUCCESS) {
        /*
          So, we have one operation to go back to tabs stack, and then another one to change tab to receiveTabRoute.

          Since we're using createTabNavigator inside inside of createStackNavigator
          and are then going to a different createStackNavigator, we use the goBack method and then the navigate method.

          For this to work, we are also passing navigator of tabs via state params and then navigating tabs stack as well.
        */
        this.props.navigation.goBack(null)
        this.props.navigation.state &&
          this.props.navigation.state.params.navigate(receiveTabRoute)
      } else if (this.props.tokenSentStatus === STORE_STATUS.ERROR) {
        this.setState({
          credentialOfferModalStatus:
            CREDENTIAL_OFFER_MODAL_STATUS.TOKEN_SENT_FAIL,
        })
      }
    }
  }

  componentDidMount() {
    this.props.navigation.setParams({
      onSendTokens: this.onSendTokens,
    })
  }

  onTextChange = (text: string, name: string) => {
    this.paymentData[name] = text
    if (this.paymentData['paymentTo'].length) {
      this.setState({
        showPaymentAddress: true,
      })
    } else {
      this.setState({
        showPaymentAddress: false,
      })
    }
  }

  onTokenSentFailedClose = () => {
    this.setState({
      credentialOfferModalStatus: CREDENTIAL_OFFER_MODAL_STATUS.NONE,
    })
  }

  throttledAsyncValidationFunction = () => {
    let status = ''
    if (this.paymentData['paymentTo'].length > 20) {
      status = 'SUCCESS'
      this.props.navigation.setParams({ isValid: true })
    } else if (this.paymentData['paymentTo'].length <= 0) {
      status = 'IDLE'
      this.props.navigation.setParams({ isValid: false })
    } else {
      status = 'ERROR'
      this.props.navigation.setParams({ isValid: false })
    }
    this.setState({
      isPaymentAddressValid: status,
    })
  }

  onShowTransactionFeesModal() {
    this.setState({
      credentialOfferModalStatus: CREDENTIAL_OFFER_MODAL_STATUS.LEDGER_FEES,
    })
  }

  onHideTransactionFeesModal = () => {
    this.setState({
      credentialOfferModalStatus: CREDENTIAL_OFFER_MODAL_STATUS.NONE,
    })
  }

  onStartTransfer = () => {
    this.onHideTransactionFeesModal()
    this.props.sendTokens(this.props.tokenAmount, this.paymentData.paymentTo)
  }

  onSendTokens = () => {
    this.onTokenSentFailedClose()
    if (this.state.isPaymentAddressValid === 'SUCCESS') {
      this.onShowTransactionFeesModal()
    }
  }

  render() {
    const testID = 'wallet-tab-send-details'
    const fontSize = scale(FONT_SIZE_MAPPING(this.props.tokenAmount.length))

    return (
      <Container safeArea fifth onPress={Keyboard.dismiss} testID={testID}>
        <ControlInput
          label="To"
          name="paymentTo"
          placeholder="Enter Payee wallet address"
          validation={this.throttledAsyncValidationFunction}
          onChangeText={this.onTextChange}
          isValid={this.state.isPaymentAddressValid}
          maxLength={100}
          autoCorrect={false}
          autoCapitalize="none"
          autoFocus={true}
        />
        <ControlInput
          label="For"
          name="paymentFor"
          multiline
          maxLength={80}
          placeholder="credential, gift, etc."
          onChangeText={this.onTextChange}
        />
        {this.props.tokenSentStatus === STORE_STATUS.IN_PROGRESS && (
          <CustomView center style={[styles.loaderContainer]}>
            <Loader showMessage={false} />
          </CustomView>
        )}
        <CustomView row center style={[{ width: '100%' }]} horizontalSpace>
          <CustomText
            quinaryText
            formatNumber
            transparentBg
            center
            style={[
              styles.tokenAmountText,
              {
                fontSize,
              },
            ]}
            fullWidth
          >
            {this.props.tokenAmount}
          </CustomText>
        </CustomView>
        <CustomView center>
          <CustomText transparentBg style={[styles.textSovrinTokens]}>
            SOVRIN TOKENS
          </CustomText>
        </CustomView>
        <CustomView center style={[{ marginTop: 36 }]}>
          <CustomView center style={[{ width: '65%' }]}>
            <CustomText transparentBg style={[styles.walletContextText]}>
              You are sending{' '}
              <CustomText transparentBg quinaryText>
                {formatNumbers(this.props.tokenAmount)}
              </CustomText>{' '}
              tokens to this wallet address:
            </CustomText>
          </CustomView>
        </CustomView>
        {this.state.showPaymentAddress ? (
          <CustomView center horizontalSpace>
            <CustomText
              center
              transparentBg
              primary
              style={[
                { marginTop: 10 },
                this.state.isPaymentAddressValid === 'SUCCESS'
                  ? styles.validTextColor
                  : styles.invalidTextColor,
              ]}
            >
              {this.paymentData.paymentTo}
            </CustomText>
          </CustomView>
        ) : (
          <CustomView center>
            <CustomText
              center
              transparentBg
              quinaryText
              style={[styles.walletAddressText]}
            >
              Enter wallet address above
            </CustomText>
          </CustomView>
        )}

        <CredentialOfferModal
          connectionName={this.paymentData.paymentTo}
          credentialOfferModalStatus={this.state.credentialOfferModalStatus}
          testID={`${testID}-payment-failure-modal`}
          onClose={this.onTokenSentFailedClose}
          onRetry={this.onSendTokens}
          onYes={this.onStartTransfer}
          onNo={this.onHideTransactionFeesModal}
          transferAmount={this.props.tokenAmount}
        />
      </Container>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: OFFSET_3X,
  },
  innerContainer: {
    ...Platform.select({
      ios: {
        borderBottomColor: color.bg.fifth.font.tertiary,
        borderBottomWidth: HAIRLINE_WIDTH,
      },
      android: {
        borderBottomColor: color.bg.fifth.font.secondary,
        borderBottomWidth: 1,
      },
    }),
    padding: OFFSET_2X,
  },
  message: {
    marginBottom: OFFSET_1X / 2,
  },
  headerLeft: {
    width: OFFSET_2X,
  },
  title: {
    marginTop: OFFSET_6X,
    marginBottom: OFFSET_7X,
  },
  titleText: {
    lineHeight: 28,
    letterSpacing: 0.5,
    paddingHorizontal: OFFSET_1X,
  },
  invalidTextColor: {
    color: color.bg.tenth.font.color,
  },
  validTextColor: {
    color: color.bg.eighth.color,
  },
  walletAddressText: {
    fontSize: 14,
    flexWrap: 'wrap',
    lineHeight: 16,
    marginTop: 11,
    marginBottom: 15,
  },
  walletContextText: {
    textAlign: 'center',
    fontSize: 14,
    lineHeight: 18,
    color: color.bg.seventh.font.sixth,
    letterSpacing: -0.35,
  },
  textSovrinTokens: {
    fontSize: 18,
    color: color.bg.seventh.font.seventh,
  },
  tokenAmountText: {
    marginTop: 19,
  },
  disabledText: {
    color: color.bg.eighth.disabled,
  },
  loaderContainer: {
    height: 40,
  },
})

const mapStateToProps = (state: Store) => ({
  tokenAmount: getTokenAmount(state),
  tokenSentStatus: state.wallet.payment.status,
})

const mapDispatchToProps = dispatch =>
  bindActionCreators({ sendTokens }, dispatch)

export default createStackNavigator({
  WalletTabSendDetails: {
    screen: connect(mapStateToProps, mapDispatchToProps)(WalletTabSendDetails),
  },
})
