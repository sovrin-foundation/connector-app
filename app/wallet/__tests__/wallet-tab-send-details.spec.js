// @flow
import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { WalletTabSendDetails } from '../wallet-tab-send-details'
import { getStore } from '../../../__mocks__/static-data'
import { getNavigation, smsToken } from '../../../__mocks__/static-data'
import { STORE_STATUS } from '../type-wallet'

describe('<WalletTabSendDetails />', () => {
  const store = getStore()
  let navigation = {
    ...getNavigation(),
  }

  function setup() {
    const component = renderer.create(
      <Provider store={store}>
        <WalletTabSendDetails
          tokenAmount="12345"
          tokenSentStatus={STORE_STATUS.SUCCESS}
          navigation={navigation}
          sendTokens={jest.fn()}
        />
      </Provider>
    )
    const instance = component.root.findByType(WalletTabSendDetails).instance
    return { component, instance }
  }

  it('should render properly and match the snapshot', () => {
    const { component, instance } = setup()
    instance.onTextChange('pay:sov:somepaymentaddress', 'paymentTo')
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
