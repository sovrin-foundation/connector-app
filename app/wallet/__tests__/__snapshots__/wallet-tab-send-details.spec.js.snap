// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WalletTabSendDetails /> should render properly and match the snapshot 1`] = `
<RCTSafeAreaView
  accessibilityComponentType={undefined}
  accessibilityLabel="wallet-tab-send-details-touchable"
  accessibilityTraits={undefined}
  accessible={false}
  fifth={true}
  hitSlop={
    Object {
      "bottom": 20,
      "left": 20,
      "right": 20,
      "top": 20,
    }
  }
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  safeArea={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="wallet-tab-send-details-touchable"
>
  <View
    accessibilityLabel="To-mainview-wallet-send-details"
    accessible={true}
    style={
      Object {
        "borderBottomColor": "#EAEAEA",
        "borderBottomWidth": 1,
      }
    }
    testID="To-mainview-wallet-send-details"
  >
    <View
      accessibilityLabel="To-subview-wallet-send-details"
      accessible={true}
      center={true}
      fifth={true}
      horizontalSpace={true}
      row={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 15,
        }
      }
      testID="To-subview-wallet-send-details"
    >
      <View
        fifth={true}
        style={
          Object {
            "alignItems": "flex-end",
            "backgroundColor": "#ffffff",
            "paddingVertical": 25,
            "width": 35,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "fontFamily": "Lato",
              },
              Object {
                "backgroundColor": "#ffffff",
                "color": "#333333",
              },
              Object {
                "fontWeight": "600",
              },
              null,
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": "rgba(237, 156, 46, 1)",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              Object {
                "lineHeight": 20,
              },
            ]
          }
        >
          To:
        </Text>
      </View>
      <View
        accessibilityLabel="To-textinput-wallet-send-details"
        accessible={true}
        fifth={true}
        horizontalSpace={true}
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#ffffff",
            "paddingHorizontal": 15,
            "width": "83%",
          }
        }
        testID="To-textinput-wallet-send-details"
      >
        <TextInput
          accessibilityLabel="To-token-send-details-label"
          accessible={true}
          autoCapitalize="none"
          autoCorrect={false}
          autoFocus={true}
          maxLength={100}
          multiline={false}
          onChangeText={[Function]}
          placeholder="Enter Payee wallet address"
          style={
            Array [
              Object {
                "fontSize": 17,
                "fontStyle": "italic",
                "lineHeight": 20,
                "width": "100%",
              },
            ]
          }
          testID="To-token-send-details-label"
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={Object {}}
      />
    </View>
  </View>
  <View
    accessibilityLabel="For-mainview-wallet-send-details"
    accessible={true}
    style={
      Object {
        "borderBottomColor": "#EAEAEA",
        "borderBottomWidth": 1,
      }
    }
    testID="For-mainview-wallet-send-details"
  >
    <View
      accessibilityLabel="For-subview-wallet-send-details"
      accessible={true}
      center={true}
      fifth={true}
      horizontalSpace={true}
      row={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingHorizontal": 15,
        }
      }
      testID="For-subview-wallet-send-details"
    >
      <View
        fifth={true}
        style={
          Object {
            "alignItems": "flex-end",
            "backgroundColor": "#ffffff",
            "paddingVertical": 25,
            "width": 35,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "fontFamily": "Lato",
              },
              Object {
                "backgroundColor": "#ffffff",
                "color": "#333333",
              },
              Object {
                "fontWeight": "600",
              },
              null,
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": "rgba(237, 156, 46, 1)",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
              Object {
                "lineHeight": 20,
              },
            ]
          }
        >
          For:
        </Text>
      </View>
      <View
        accessibilityLabel="For-textinput-wallet-send-details"
        accessible={true}
        fifth={true}
        horizontalSpace={true}
        style={
          Object {
            "alignItems": "flex-start",
            "backgroundColor": "#ffffff",
            "paddingHorizontal": 15,
            "width": "83%",
          }
        }
        testID="For-textinput-wallet-send-details"
      >
        <TextInput
          accessibilityLabel="For-token-send-details-label"
          accessible={true}
          maxLength={80}
          multiline={true}
          onChangeText={[Function]}
          placeholder="credential, gift, etc."
          style={
            Array [
              Object {
                "fontSize": 17,
                "fontStyle": "italic",
                "lineHeight": 20,
                "width": "100%",
              },
            ]
          }
          testID="For-token-send-details-label"
          underlineColorAndroid="transparent"
        />
      </View>
      <View
        style={Object {}}
      />
    </View>
  </View>
  <View
    center={true}
    horizontalSpace={true}
    row={true}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 15,
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "fontFamily": "Lato",
          },
          Object {
            "backgroundColor": "#333333",
            "color": "#ffffff",
          },
          null,
          Object {
            "textAlign": "center",
          },
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "color": "rgba(237, 156, 46, 1)",
          },
          null,
          null,
          null,
          Object {
            "width": "100%",
          },
          null,
          null,
          null,
          Object {
            "marginTop": 19,
          },
          Object {
            "fontSize": 54.857142857142854,
          },
        ]
      }
    >
      12,345
    </Text>
  </View>
  <View
    center={true}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "fontFamily": "Lato",
          },
          Object {
            "backgroundColor": "#333333",
            "color": "#ffffff",
          },
          null,
          null,
          Object {
            "backgroundColor": "transparent",
          },
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          Object {
            "color": "#9B9B9B",
            "fontSize": 18,
          },
        ]
      }
    >
      SOVRIN TOKENS
    </Text>
  </View>
  <View
    center={true}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 36,
      }
    }
  >
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "width": "65%",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "Lato",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "color": "#6D6D6D",
              "fontSize": 14,
              "letterSpacing": -0.35,
              "lineHeight": 18,
              "textAlign": "center",
            },
          ]
        }
      >
        You are sending
         
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontSize": 17,
              },
              Object {
                "fontFamily": "Lato",
              },
              Object {
                "backgroundColor": "#333333",
                "color": "#ffffff",
              },
              null,
              null,
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "color": "rgba(237, 156, 46, 1)",
              },
              null,
              null,
              null,
              null,
              null,
              null,
              null,
            ]
          }
        >
          12,345
        </Text>
         
        tokens to this wallet address:
      </Text>
    </View>
  </View>
  <View
    center={true}
    horizontalSpace={true}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingHorizontal": 15,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "fontFamily": "Lato",
          },
          Object {
            "backgroundColor": "#333333",
            "color": "#ffffff",
          },
          null,
          Object {
            "textAlign": "center",
          },
          Object {
            "backgroundColor": "transparent",
          },
          null,
          null,
          null,
          Object {
            "color": "#333333",
          },
          null,
          null,
          null,
          null,
          Object {
            "marginTop": 10,
          },
          Object {
            "color": "#d0021b",
          },
        ]
      }
    >
      pay:sov:somepaymentaddress
    </Text>
  </View>
  <Modal
    animationIn="zoomIn"
    animationOut="zoomOut"
    animationOutTiming={100}
    backdropColor="#f0f0f0"
    hideModalContentWhileAnimating={true}
    isVisible={false}
    onBackButtonPress={[Function]}
    onBackdropPress={[Function]}
    onModalHide={undefined}
    style={
      Array [
        Object {},
      ]
    }
    testID="credential-offer-modal"
    useNativeDriver={true}
  />
</RCTSafeAreaView>
`;
